stacks_queues/queue.py

```diff
@@ -11,52 +16,137 @@ class Queue:
+        self.front = 0 # ~ head
+        self.rear = 0 # ~ tail
+        self.size = 0 # number of elements currently in the queue
+
+    def isfull_helper(self):
+        """ Returns True if queque is full
+        """
+        return self.size == self.buffer_size
 
     def enqueue(self, element):
         """ Adds an element to the Queue
             Raises a QueueFullException if all elements
             In the store are occupied
             returns None
         """
-        pass
+        # Check if queque is full using helper method:
+        if (self.isfull_helper()): 
+            # If full, raise error:
+            raise QueueFullException
+
+        # If queque is not full, it's safe to insert an element
+        # The current tail index is also the index where we want to insert the element, 
+        # since we're always adding to the back of the queue. So we need to store the 
+        # element at the store position of the tail:
+        self.store[self.rear] = element
+
+        # Then, need to advance the tail index:
+        self.rear = (self.rear + 1) % self.buffer_size
+
+        # Also need to increment the size to account for the newly added element:
+        self.size += 1
 
+
+    # ❗️❗️❗️❗️ According to the tests, this method should return a value, not None. 
     def dequeue(self):
         """ Removes an element from the Queue
             Raises a QueueEmptyException if 
             The Queue is empty.
-            returns None
+            returns None 
         """
-        pass
+        # Check if queque is empty, and returns none if so:
+        if (self.size == 0):
+            raise QueueEmptyException
+            
+        value = self.store[self.front]
+        self.front = (self.front + 1) % self.buffer_size
+
+        # Then need to decrement the size to account for the element that was removed:
+        self.size -= 1
+
+        return value 
+
 
-    def front(self):
+    def get_front(self):
         """ Returns an element from the front
             of the Queue and None if the Queue
             is empty.  Does not remove anything.
         """
-        pass
+        if self.size == 0:
+            return None
+
+        return self.store[self.front]
         
 
     def size(self):
         """ Returns the number of elements in
             The Queue
         """
-        pass
+        return self.size 
 
     def empty(self):
         """ Returns True if the Queue is empty
             And False otherwise.
         """
-        pass
+        return (self.size == 0)
 
     def __str__(self):
         """ Returns the Queue in String form like:
             [3, 4, 7]
             Starting with the front of the Queue and
             ending with the rear of the Queue.
         """
-        pass
+        if self.size == 0:
+            return "[]"
+
+        store_string = "["
+        for elem in range(self.front, (self.front + self.size)):
+            store_string += (f"{self.store[elem]}, ")
```

Comment:
I'm pretty sure this can make the index go beyond the bounds of the array

This might fix it.

```suggestion
        for elem in range(self.front, (self.front + self.size)):
            store_string += (f"{self.store[elem % self.buffer_size]}, ")
```